impeller/renderer/backend/vulkan/device_holder.h

```diff
@@ -0,0 +1,17 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#pragma once
+
+#include "impeller/renderer/backend/vulkan/vk.h"
+
+namespace impeller {
+
+class DeviceHolder {
+ public:
+  virtual ~DeviceHolder() = default;
+  virtual const vk::Device* GetDevice() const = 0;
```

Comment:
`DeviceHolder::GetDevice()` should return a `vk::Device` (instead of a `vk::Device*`).  And code that uses a `vk::Device` should pass instances (or const references) rather than pointers.

Passing handle wrappers like `vk::Device` by value is consistent with how the vulkan-hpp API itself uses these objects.

IIUC the intent of this change was to make the C++ type reflect the ownership semantics.  But developers of this subsystem are going to need to understand the behavior of the Khronos Vulkan C++ wrappers and that a `vk::Device` instance is equivalent to a Vulkan device handle.

The `const vk::Device*` inaccurately implies that the `vk::Device` is nullable, and creates the potential for someone to be holding onto a `vk::Device*` pointing to a C++ object that has been deleted.

Reply:
The reason I went with this approach is that the whole reason we got into this mess is that people were copying the `vk::Device` and not realize what they actually had is a raw pointer that may be invalid.  By passing around a `std::weak_ptr<DeviceHolder>` which returns a pointer to the device, my hope is that this would make that error much less likely.

I can make DeviceHolder return a `const vk::Device&` but that is going to make it so much easier for someone to copy the `vk::Device` without realizing the semantics of it since `vk::Device` has a copy constructor.

Alternatively, I might be able to do `virtual const vk::UniqueDevice& GetDevice() const = 0`. Which would give you the guarantees you want and make accidentally messing things up less likely.

Here's where I want to be: no one has an lvalue to a `vk::Device`.

Reply:
If other modules in this subsystem are interacting directly with Vulkan devices, then they will need to understand how Vulkan devices work and how their lifetime is managed.  For now I think the most idiomatic option is having `DeviceHolder::GetDevice` return a `const vk::Device&` and establishing a convention that anyone who may need the Vulkan device holds a `weak_ptr<DeviceHolder>` and should never hold a raw `vk::Device` instance.

The type system will not be able to prevent all forms of misuse as long as the other modules require direct access to the `vk::Device`.  If the convention above is insufficient, then we should hide all usage of the `vk::Device` behind some other interface.

Reply:
I think it's a loaded gun allowing `vk::Device&`.  I'd be down with making our own wrapper in the future.  For now I put in the references as requested.